### PR TITLE
 Refine document and scripts of CTC model.

### LINKED ISSUE #764: 能否完善fluid下的OCR的文档
在这个OCR例子中：https://github.com/PaddlePaddle/models/tree/develop/fluid/ocr_recognition
文档较少，能否完善一下文档，比如说说每个文件的作用。

### PR DESCRIPTION
**Pull request by wanghaoshuang**
fix #764
1. Add document.
2. Add arguments for saving model and init model.
3. Refine inference.py and eval.py.
4. Make ctc_reader.py support for custom data.

### CODE REVIEW

File: fluid/ocr_recognition/README.md [hunk 1]
```diff
@@ -1,4 +1,192 @@
...
+
+**- -train_images :** 存放训练集图片的路径，如果设置为None，ctc_reader会自动下载使用默认数据集。如果使用自己的数据进行训练，需要修改该选项。默认为None。
+
+**- -train_list :** 存放训练集图片信息的list文件，如果设置为None，ctc_reader会自动下载使用默认数据集。如果使用自己的数据进行训练，需要修改该选项。默认为None。
+
+**- -test_images :** 存放测试集图片的路径，如果设置为None，ctc_reader会自动下载使用默认数据集。如果使用自己的数据进行测试，需要修改该选项。默认为None。
+
+**- -test_list :** 存放测试集图片信息的list文件，如果设置为None，ctc_reader会自动下载使用默认数据集。如果使用自己的数据进行测试，需要修改该选项。默认为None。
+
+**- -num_classes :** 字符集的大小。如果设置为None, 则使用ctc_reader提供的字符集大小。如果使用自己的数据进行训练，需要修改该选项。默认为None.
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 这里觉得不用每个参数都解释，以后变动就得改。可以train.py里的参数注释写好点，告诉用户: `python train.py --help`查看使用方法。

File: fluid/ocr_recognition/README.md [hunk 2]
```diff
@@ -1,4 +1,192 @@
...
+
+其中，options支持配置以下预测相关的参数：
+
+**--model_path :**  用来做预测的模型文件。如果模型是以单个文件存储的，这里需要指定具体文件的路径；如果模型是以多个文件存储的，这里只需指定多个文件所在文件夹的路径。为必设置选项。
+
+**--input_images_dir :** 存放待预测图片的文件夹路径。如果设置为None, 则使用ctc_reader提供的默认数据。默认为None.
+
+**--input_images_list :**  存放待预测图片信息的list文件的路径。如果设置为None, 则使用ctc_reader提供的默认数据。默认为None.
+
+**--device DEVICE :** 设备ID。设置为-1，运行在CPU上；设置为0，运行在GPU上。默认为0。
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 同上~

File: fluid/ocr_recognition/README.md [hunk 3]
```diff
@@ -1,4 +1,192 @@
...
+
+预测结果会print到标准输出。
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 关于OCR的预测，最好输入一张图片，输出一个文本，结果显示出来。

File: fluid/ocr_recognition/README.md [hunk 4]
```diff
@@ -1,4 +1,192 @@
...
+
+其中，options支持配置以下评估相关的参数：
+
+**--model_path :**  待评估模型的文件路径。如果模型是以单个文件存储的，这里需要指定具体文件的路径；如果模型是以多个文件存储的，这里只需指定多个文件所在文件夹的路径。为必设置选项。
+
+**--input_images_dir :** 存放待评估图片的文件夹路径。如果设置为None, 则使用ctc_reader提供的默认数据。默认为None.
+
+**--input_images_list :**  存放待评估图片信息的list文件的路径。如果设置为None, 则使用ctc_reader提供的默认数据。默认为None.
+
+**--device DEVICE :** 设备ID。设置为-1，运行在CPU上；设置为0，运行在GPU上。默认为0。
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 同上，Evaluation可以放在Inference前面？

File: fluid/ocr_recognition/README.md (hunk 1 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done.

File: fluid/ocr_recognition/README.md (hunk 2 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done.

File: fluid/ocr_recognition/README.md (hunk 4 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done.

File: fluid/ocr_recognition/README.md (hunk 3 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: 直接输出文本有点困难，还需要用户给一个字典。
现在是用户可以输入一个图片路径，然后即时输出一个indexes 序列结果.
```
python inference.py --model_path models/model_00044_15000
-----------  Configuration Arguments -----------
device: 0
input_images_dir: None
input_images_list: None
model_path: models/model_00044_15000
------------------------------------------------
Init model from: models/model_00044_15000.
Please input the path of image: data/test_images/00008_4700.jpg
result: [6514 5919 3415  173]
Please input the path of image:
```

File: fluid/ocr_recognition/README.md [hunk 5]
```diff
@@ -1,4 +1,170 @@
-# OCR Model
+﻿
+[toc]
 
-This model built with paddle fluid is still under active development and is not
-the final version. We welcome feedbacks.
+运行本目录下的程序示例需要使用PaddlePaddle v0.11.0 版本。如果您的PaddlePaddle安装版本低于此要求，请按照安装文档中的说明更新PaddlePaddle安装版本。
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 这里要求需要更新，让使用Develop最新版本吧，后续稳定了，我们在更改。

File: fluid/ocr_recognition/README.md [hunk 6]
```diff
@@ -1,4 +1,170 @@
...
+```
+
+执行`python ctc_train.py --help`可查看更多使用方式和参数详细说明。
+
+图2为使用默认参数和默认数据集训练的收敛曲线，其中横坐标轴为训练pass数，纵轴为在测试集上的sequence_error.
+
+<p align="center">
+<img src="images/train.jpg" width="620" hspace='10'/> <br/>
+<strong>图 2</strong>
+</p>
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md:130]: https://github.com/wanghaoshuang/models/tree/ctc_doc/fluid/ocr_recognition  图2没显示出来。

File: fluid/ocr_recognition/README.md [hunk 7]
```diff
@@ -1,4 +1,170 @@
...
+
+
+
+### 1.3 Evaluate
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 这个是中文文档，请标题也使用中文。

File: fluid/ocr_recognition/README.md [hunk 8]
```diff
@@ -1,4 +1,170 @@
...
+### 1.4 Inference
+
+Read image path from stdin and inference：
+
+```
+env CUDA_VISIBLE_DEVICE=0 python inference.py \
+    --model_path models/model_00044_15000
+```
+
+Read image path from list file and inference：
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 语言统一使用中文。

File: fluid/ocr_recognition/README.md [hunk 9]
```diff
@@ -1,4 +1,170 @@
...
+
+```
+env CUDA_VISIBLE_DEVICE=0 python inference.py \
+    --model_path=models/model_00044_15000 \
+    --input_images_list="data/test.list"
+```
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 请给出预测结果示例。

File: fluid/ocr_recognition/README.md [hunk 10]
```diff
@@ -1,4 +1,170 @@
...
+
+第三种：从stdin读入一张图片的path，然后进行一次inference.
+
+#### 1.2 训练
+
+Train on one GPU with defalut trainning data:
+
+```
+env CUDA_VISIABLE_DEVICES=0 python ctc_train.py \
+    --device=0 \
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 代码里可以去掉`device`，换成`use_gpu`?

File: fluid/ocr_recognition/README.md [hunk 11]
```diff
@@ -1,4 +1,170 @@
...
+    --parallel=False \
+    --batch_size=32
+```
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md]: 这些参数都是默认的话，写成下面这样？ 方便用户直接粘贴赋值。
```
env CUDA_VISIABLE_DEVICES=0 python ctc_train.py
```

File: fluid/ocr_recognition/ctc_train.py [hunk 12]
```diff
@@ -1,61 +1,83 @@
...
+add_arg('pass_num',          int,   100,        "Number of training epochs.")
+add_arg('log_period',        int,   1000,       "Log period.")
+add_arg('save_model_period', int,   15000,      "Save model period. '-1' means never saving the model.")
+add_arg('eval_period',       int,   15000,      "Evaluate period. '-1' means never evaluating the model.")
+add_arg('save_model_dir',    str,   "./models", "The directory the model to be saved to.")
+add_arg('init_model',        str,   None,       "The init model file of directory.")
+add_arg('learning_rate',     float, 1.0e-3,    "Learning rate.")
+add_arg('l2',                float, 0.0004,    "L2 regularizer.")
+add_arg('max_clip',          float, 10.0,      "Max clip threshold.")
+add_arg('min_clip',          float, -10.0,     "Min clip threshold.")
```
Comment (qingqing01) [inline fluid/ocr_recognition/ctc_train.py]: max_clip/min_clip用了吗？ 感觉尽量减少参数。

File: fluid/ocr_recognition/ctc_train.py [hunk 13]
```diff
@@ -1,61 +1,83 @@
...
+add_arg('train_images',      str,   None,    "The directory of training images."
+        "None means using the default training images of reader.")
+add_arg('train_list',        str,   None,    "The list file of training images."
+        "None means using the default train_list file of reader.")
+add_arg('test_images',      str,    None,    "The directory of training images."
+        "None means using the default test images of reader.")
+add_arg('test_list',        str,    None,   "The list file of training images."
+        "None means using the default test_list file of reader.")
+add_arg('num_classes',      int,    None,      "The number of classes."
+        "None means using the default num_classes from reader.")
```
Comment (qingqing01) [inline fluid/ocr_recognition/ctc_train.py:40]: 觉得用户可配置的参数是做实验【频繁】使用的参数。 这里的参数太多了。

File: fluid/ocr_recognition/README.md (hunk 5 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done. Thx.

File: fluid/ocr_recognition/README.md (hunk 7 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done. Thx.

File: fluid/ocr_recognition/README.md (hunk 8 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done. Thx.

File: fluid/ocr_recognition/README.md (hunk 9 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done. Thx.

File: fluid/ocr_recognition/README.md (hunk 10 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done. Thx.

File: fluid/ocr_recognition/README.md (hunk 11 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md]: Done. Thx.

File: fluid/ocr_recognition/ctc_train.py (hunk 12 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/ctc_train.py]: Removed.

File: fluid/ocr_recognition/README.md (hunk 6 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/README.md:130]: Fixed. Thx.

File: fluid/ocr_recognition/ctc_train.py (hunk 13 above)
Comment (wanghaoshuang) [reply fluid/ocr_recognition/ctc_train.py:40]: 感觉还好吧，有些参数虽然不经常调整，但是也是必不可少的。

File: fluid/ocr_recognition/README.md [hunk 14]
```diff
@@ -1,4 +1,179 @@
-# OCR Model
+﻿
+[toc]
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md:2]: https://github.com/wanghaoshuang/models/tree/ctc_doc/fluid/ocr_recognition

toc 没显示出来。

File: fluid/ocr_recognition/README.md [hunk 15]
```diff
@@ -1,4 +1,179 @@
...
 
-This model built with paddle fluid is still under active development and is not
-the final version. We welcome feedbacks.
+运行本目录下的程序示例需要使用PaddlePaddle develop最新版本。如果您的PaddlePaddle安装版本低于此要求，请按照安装文档中的说明更新PaddlePaddle安装版本。
+
+# Optical Character Recognition
+
+这里将介绍如何在PaddlePaddle fluid下使用CRNN-CTC 和 CRNN-Attention模型对图片中的文字内容进行识别。
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md:8]: fluid -> Fluid

File: fluid/ocr_recognition/README.md [hunk 16]
```diff
@@ -1,4 +1,179 @@
...
+
+## 1. CRNN-CTC
+
+本章的任务是识别含有单行汉语字符图片，首先采用卷积将图片转为`features map`, 然后使用`im2sequence op`将`features map`转为`sequence`，经过`双向GRU RNN`得到每个step的汉语字符的概率分布。训练过程选用的损失函数为CTC loss，最终的评估指标为`instance error rate`。
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md:12]: features map换成中文吧，叫特征图。
sequence-> 序列。

> 经过`双向GRU RNN`得到每个step的汉语字符的概率分布 

实际模型里并没有得到概率分布。 

通过`双向GRU`学习到序列特征。

第一出现的CTC地方，需要中文。

`instance error rate`: 需要解释明白， 可以写成： 样本级别的错误率。

File: fluid/ocr_recognition/README.md [hunk 17]
```diff
@@ -1,4 +1,179 @@
...
+
+本路径下各个文件的作用如下：
+
+- **ctc_reader.py :** 下载、读取、处理数据。提供方法`train()` 和 `test()` 分别产生训练集和测试集的数据迭代器。
+- **crnn_ctc_model.py :** 在该脚本中定义了训练网络、预测网络和evaluate网络。
+- **ctc_train.py :** 用于模型的训练，可通过命令`python train.py --help` 获得使用方法。
+- **inference.py :** 加载训练好的模型文件，对新数据进行预测。可通过命令`python inference.py --help` 获得使用方法。
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md:19]: `inference.py` -> `infer.py`

File: fluid/ocr_recognition/README.md [hunk 18]
```diff
@@ -1,4 +1,179 @@
...
+#### 1.1.1 数据格式
+
+我们使用的训练和测试数据如`图1`所示，每张图片包含单行不定长的中文字符串，这些图片都是经过检测算法进行预框选处理的。
+
+<p align="center">
+<img src="images/demo.jpg" width="620" hspace='10'/> <br/>
+<strong>图 1</strong>
+</p>
+
+在训练集中，每张图片对应的label是由若干数字组成的sequence。 Sequence中的每个数字表示一个字符在字典中的index。 `图1` 对应的label如下所示：
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md:37]: > 每张图片对应的label是由若干数字组成的sequence。Sequence中的每个数字表示一个字符在字典中的index。

每张图片对应的label是汉字在词典中的索引。

File: fluid/ocr_recognition/README.md [hunk 19]
```diff
@@ -1,4 +1,179 @@
...
+-----------  Configuration Arguments -----------
+use_gpu: True
+input_images_dir: None
+input_images_list: None
+model_path: /home/work/models/fluid/ocr_recognition/models/model_00052_15000
+------------------------------------------------
+Init model from: /home/work/models/fluid/ocr_recognition/models/model_00052_15000.
+Please input the path of image: /home/work/models/fluid/ocr_recognition/data/test_images/00001_0060.jpg
+result: [3298 2371 4233 6514 2378 3298 2363]
+Please input the path of image: /home/work/models/fluid/ocr_recognition/data/test_images/00001_0429.jpg
```
Comment (qingqing01) [inline fluid/ocr_recognition/README.md:169]: >  /home/work/models/fluid/ocr_recognition/data/test_images/00001_0429.jpg
这样的路径在文档中，对用户不友好。

这里可以使用图1的图片吗？ 输出结果可以换成在词典中转换之后的汉字吗？

File: fluid/ocr_recognition/README.md (hunk 6 above)
Comment (qingqing01) [reply fluid/ocr_recognition/README.md:130]: 1. 给出的图的同时，解释下图的意思， 以及说下seq error是多少。 有train的seq error吗？如果有画两条？
2. 是否需要给出train和test的cost图？

**Review by qingqing01 — APPROVED** [review]
I approve the PR. But some comments need to fix in next PR.
